File: backend/app/application/schemas/llm.py
"""
LLM Configuration Schemas
Request and response schemas for LLM model configuration
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Union
from datetime import datetime


# ============ Response Schemas ============

class LLMFactoryResponse(BaseModel):
    """LLM 厂商响应"""
    name: str
    logo: str = ""
    tags: str
    rank: int = 0
    status: str = "1"
    model_types: List[str] = []

    class Config:
        from_attributes = True


class LLMModelResponse(BaseModel):
    """LLM 模型响应"""
    fid: str
    llm_name: str
    model_type: str
    max_tokens: int = 0
    available: bool = False
    status: str = "1"
    tags: str = ""
    is_tools: bool = False

    class Config:
        from_attributes = True


class TenantLLMDetail(BaseModel):
    """租户 LLM 详情"""
    type: str
    name: str
    used_token: int = 0
    api_base: str = ""
    max_tokens: int = 8192
    status: str = "1"


class MyLlmValue(BaseModel):
    """我的 LLM 值"""
    tags: str
    llm: List[TenantLLMDetail]


class TenantInfoResponse(BaseModel):
    """租户信息响应"""
    tenant_id: str
    name: Optional[str] = None
    llm_id: Optional[str] = None
    embd_id: Optional[str] = None
    asr_id: Optional[str] = None
    img2txt_id: Optional[str] = None
    rerank_id: Optional[str] = None
    tts_id: Optional[str] = None

    class Config:
        from_attributes = True


# ============ Request Schemas ============

class SetApiKeyRequest(BaseModel):
    """设置 API Key 请求"""
    llm_factory: str = Field(..., description="厂商名称")
    api_key: str = Field(..., description="API Key")
    base_url: Optional[str] = Field(None, description="API Base URL")
    model_type: Optional[str] = Field(None, description="模型类型")
    llm_name: Optional[str] = Field(None, description="模型名称")


class AddLlmRequest(BaseModel):
    """添加 LLM 请求"""
    llm_factory: str = Field(..., description="厂商名称")
    llm_name: str = Field(..., description="模型名称")
    model_type: str = Field(..., description="模型类型: chat, embedding, image2text, speech2text, rerank, tts, ocr")
    api_base: Optional[str] = Field(None, description="API Base URL")
    api_key: Union[str, Dict[str, Any]] = Field(..., description="API Key 或配置对象")
    max_tokens: Optional[int] = Field(None, description="最大 Token 数")

    # 特殊厂商字段
    # Azure OpenAI
    api_version: Optional[str] = None
    # VolcEngine
    ark_api_key: Optional[str] = None
    endpoint_id: Optional[str] = None
    # Bedrock
    auth_mode: Optional[str] = None
    bedrock_ak: Optional[str] = None
    bedrock_sk: Optional[str] = None
    bedrock_region: Optional[str] = None
    aws_role_arn: Optional[str] = None
    # Google Cloud
    google_project_id: Optional[str] = None
    google_region: Optional[str] = None
    google_service_account_key: Optional[str] = None
    # OpenRouter
    provider_order: Optional[str] = None
    # XunFei Spark
    spark_api_password: Optional[str] = None
    spark_app_id: Optional[str] = None
    spark_api_secret: Optional[str] = None
    spark_api_key: Optional[str] = None
    # Baidu Yiyan
    yiyan_ak: Optional[str] = None
    yiyan_sk: Optional[str] = None
    # Fish Audio
    fish_audio_ak: Optional[str] = None
    fish_audio_refid: Optional[str] = None
    # Tencent Hunyuan
    hunyuan_sid: Optional[str] = None
    hunyuan_sk: Optional[str] = None
    # Tencent Cloud
    tencent_cloud_sid: Optional[str] = None
    tencent_cloud_sk: Optional[str] = None
    # MinerU
    mineru_backend: Optional[str] = None
    mineru_server_url: Optional[str] = None
    mineru_delete_output: Optional[bool] = None


class DeleteLlmRequest(BaseModel):
    """删除 LLM 请求"""
    llm_factory: str = Field(..., description="厂商名称")
    llm_name: Optional[str] = Field(None, description="模型名称")


class EnableLlmRequest(BaseModel):
    """启用/禁用 LLM 请求"""
    llm_factory: str = Field(..., description="厂商名称")
    llm_name: str = Field(..., description="模型名称")
    status: str = Field(..., description="状态: 1=启用, 0=禁用")


class UpdateTenantRequest(BaseModel):
    """更新租户请求"""
    tenant_id: str = Field(..., description="租户 ID")
    name: Optional[str] = None
    llm_id: Optional[str] = None
    embd_id: Optional[str] = None
    asr_id: Optional[str] = None
    img2txt_id: Optional[str] = None
    rerank_id: Optional[str] = None
    tts_id: Optional[str] = None


# ============ Common Response Schemas ============

class ApiResponse(BaseModel):
    """API 响应基类"""
    code: int = 0
    data: Any = None
    message: Optional[str] = None


class LlmListResponse(BaseModel):
    """LLM 列表响应"""
    OpenAI: Optional[List[LLMModelResponse]] = None
    Anthropic: Optional[List[LLMModelResponse]] = None
    Azure_OpenAI: Optional[List[LLMModelResponse]] = None
    Google_Cloud: Optional[List[LLMModelResponse]] = None
    Ollama: Optional[List[LLMModelResponse]] = None
    Xinference: Optional[List[LLMModelResponse]] = None
    LocalAI: Optional[List[LLMModelResponse]] = None
    LM_Studio: Optional[List[LLMModelResponse]] = None
    HuggingFace: Optional[List[LLMModelResponse]] = None
    VolcEngine: Optional[List[LLMModelResponse]] = None
    Tongyi_Qianwen: Optional[List[LLMModelResponse]] = None
    Moonshot: Optional[List[LLMModelResponse]] = None
    ZHIPU_AI: Optional[List[LLMModelResponse]] = None
    DeepSeek: Optional[List[LLMModelResponse]] = None
    Minimax: Optional[List[LLMModelResponse]] = None
    Tencent_Hunyuan: Optional[List[LLMModelResponse]] = None
    Tencent_Cloud: Optional[List[LLMModelResponse]] = None
    XunFei_Spark: Optional[List[LLMModelResponse]] = None
    BaiduYiyan: Optional[List[LLMModelResponse]] = None
    Fish_Audio: Optional[List[LLMModelResponse]] = None
    Youdao: Optional[List[LLMModelResponse]] = None
    BAAI: Optional[List[LLMModelResponse]] = None
    OpenRouter: Optional[List[LLMModelResponse]] = None
    MinerU: Optional[List[LLMModelResponse]] = None
    Bedrock: Optional[List[LLMModelResponse]] = None
    Gemini: Optional[List[LLMModelResponse]] = None
    Mistral: Optional[List[LLMModelResponse]] = None


class MyLlmResponse(BaseModel):
    """我的 LLM 响应"""
    OpenAI: Optional[MyLlmValue] = None
    Anthropic: Optional[MyLlmValue] = None
    Azure_OpenAI: Optional[MyLlmValue] = None
    Google_Cloud: Optional[MyLlmValue] = None
    Ollama: Optional[MyLlmValue] = None
    Xinference: Optional[MyLlmValue] = None
    LocalAI: Optional[MyLlmValue] = None
    LM_Studio: Optional[MyLlmValue] = None
    HuggingFace: Optional[MyLlmValue] = None
    VolcEngine: Optional[MyLlmValue] = None
    Tongyi_Qianwen: Optional[MyLlmValue] = None
    Moonshot: Optional[MyLlmValue] = None
    ZHIPU_AI: Optional[MyLlmValue] = None
    DeepSeek: Optional[MyLlmValue] = None
    Minimax: Optional[MyLlmValue] = None
    Tencent_Hunyuan: Optional[MyLlmValue] = None
    Tencent_Cloud: Optional[MyLlmValue] = None
    XunFei_Spark: Optional[MyLlmValue] = None
    BaiduYiyan: Optional[MyLlmValue] = None
    Fish_Audio: Optional[MyLlmValue] = None
    Youdao: Optional[MyLlmValue] = None
    BAAI: Optional[MyLlmValue] = None
    OpenRouter: Optional[MyLlmValue] = None
    MinerU: Optional[MyLlmValue] = None
    Bedrock: Optional[MyLlmValue] = None
    Gemini: Optional[MyLlmValue] = None
    Mistral: Optional[MyLlmValue] = None

File: frontend/src/constants/llm.ts
/**
 * LLM Provider Factory Constants
 * Based on RAGFlow LLM configuration
 */

export enum LLMFactory {
  // OpenAI Family
  OpenAI = 'OpenAI',
  AzureOpenAI = 'Azure-OpenAI',
  OpenAiAPICompatible = 'OpenAI-API-Compatible',

  // Chinese Providers
  TongYiQianWen = 'Tongyi-Qianwen',
  ZhipuAI = 'ZHIPU-AI',
  WenXinYiYan = '文心一言',
  DeepSeek = 'DeepSeek',
  Moonshot = 'Moonshot',
  BaiChuan = 'BaiChuan',
  MiniMax = 'MiniMax',
  StepFun = 'StepFun',
  TencentHunYuan = 'Tencent Hunyuan',
  TencentCloud = 'Tencent Cloud',
  XunFeiSpark = 'XunFei Spark',
  BaiduYiYan = 'BaiduYiyan',

  // International Providers
  Anthropic = 'Anthropic',
  Cohere = 'Cohere',
  Mistral = 'Mistral',
  Groq = 'Groq',
  Perplexity = 'Perplexity',
  TogetherAI = 'TogetherAI',
  LeptonAI = 'LeptonAI',
  NovitaAI = 'NovitaAI',
  Upstage = 'Upstage',

  // Cloud Platforms
  GoogleCloud = 'Google Cloud',
  Gemini = 'Gemini',
  Bedrock = 'Bedrock',
  VolcEngine = 'VolcEngine',
  NVIDIA = 'NVIDIA',

  // Open Source / Local
  Ollama = 'Ollama',
  Xinference = 'Xinference',
  LocalAI = 'LocalAI',
  LMStudio = 'LM-Studio',
  ModelScope = 'ModelScope',
  HuggingFace = 'HuggingFace',
  GPUStack = 'GPUStack',
  VLLM = 'VLLM',

  // Routing / Aggregation
  OpenRouter = 'OpenRouter',
  Replicate = 'Replicate',
  PPIO = 'PPIO',
  SILICONFLOW = 'SILICONFLOW',
  PerfXCloud = 'PerfXCloud',

  // Specialized
  FishAudio = 'Fish Audio',
  Jina = 'Jina',
  VoyageAI = 'Voyage AI',
  BAAI = 'BAAI',
  NomicAI = 'nomic-ai',
  SentenceTransformers = 'sentence-transformers',

  // Other
  GiteeAI = 'GiteeAI',
  Ai302 = '302.AI',
  DeepInfra = 'DeepInfra',
  Grok = 'Grok',
  XAI = 'xAI',
  TokenPony = 'TokenPony',
  Meituan = 'Meituan',
  Longcat = 'LongCat',
  CometAPI = 'CometAPI',
  DeerAPI = 'DeerAPI',
  JiekouAI = 'Jiekou.AI',
  Builtin = 'Builtin',
  MinerU = 'MinerU',
  YouDao = 'Youdao',
}

// Icon name mapping for each LLM provider
export const IconMap: Record<LLMFactory, string> = {
  [LLMFactory.TongYiQianWen]: 'tongyi-qianwen',
  [LLMFactory.Moonshot]: 'moonshot',
  [LLMFactory.OpenAI]: 'openai',
  [LLMFactory.ZhipuAI]: 'zhipu',
  [LLMFactory.WenXinYiYan]: 'wenxin',
  [LLMFactory.Ollama]: 'ollama',
  [LLMFactory.Xinference]: 'xinference',
  [LLMFactory.ModelScope]: 'modelscope',
  [LLMFactory.DeepSeek]: 'deepseek',
  [LLMFactory.VolcEngine]: 'volcengine',
  [LLMFactory.BaiChuan]: 'baichuan',
  [LLMFactory.Jina]: 'jina',
  [LLMFactory.MiniMax]: 'MiniMax',
  [LLMFactory.Mistral]: 'mistral',
  [LLMFactory.AzureOpenAI]: 'azure',
  [LLMFactory.Bedrock]: 'bedrock',
  [LLMFactory.Gemini]: 'gemini',
  [LLMFactory.Groq]: 'groq',
  [LLMFactory.OpenRouter]: 'open-router',
  [LLMFactory.LocalAI]: 'local-ai',
  [LLMFactory.StepFun]: 'stepfun',
  [LLMFactory.NVIDIA]: 'nvidia',
  [LLMFactory.LMStudio]: 'lm-studio',
  [LLMFactory.OpenAiAPICompatible]: 'openai-api',
  [LLMFactory.Cohere]: 'cohere',
  [LLMFactory.LeptonAI]: 'lepton',
  [LLMFactory.TogetherAI]: 'together',
  [LLMFactory.PerfXCloud]: 'perfx-cloud',
  [LLMFactory.Upstage]: 'upstage',
  [LLMFactory.NovitaAI]: 'novita-ai',
  [LLMFactory.SILICONFLOW]: 'siliconflow',
  [LLMFactory.PPIO]: 'ppio',
  [LLMFactory.Replicate]: 'replicate',
  [LLMFactory.TencentHunYuan]: 'hunyuan',
  [LLMFactory.XunFeiSpark]: 'spark',
  [LLMFactory.BaiduYiYan]: 'wenxinyiyan',
  [LLMFactory.FishAudio]: 'fish-audio',
  [LLMFactory.TencentCloud]: 'tencent-cloud',
  [LLMFactory.Anthropic]: 'anthropic',
  [LLMFactory.VoyageAI]: 'voyage',
  [LLMFactory.GoogleCloud]: 'google-cloud',
  [LLMFactory.HuggingFace]: 'huggingface',
  [LLMFactory.YouDao]: 'youdao',
  [LLMFactory.BAAI]: 'baai',
  [LLMFactory.NomicAI]: 'nomic-ai',
  [LLMFactory.JinaAI]: 'jina',
  [LLMFactory.SentenceTransformers]: 'sentence-transformers',
  [LLMFactory.GPUStack]: 'gpustack',
  [LLMFactory.VLLM]: 'vllm',
  [LLMFactory.GiteeAI]: 'gitee-ai',
  [LLMFactory.Ai302]: 'ai302',
  [LLMFactory.DeepInfra]: 'deepinfra',
  [LLMFactory.Grok]: 'grok',
  [LLMFactory.XAI]: 'xai',
  [LLMFactory.TokenPony]: 'tokenpony',
  [LLMFactory.Meituan]: 'longcat',
  [LLMFactory.Longcat]: 'longcat',
  [LLMFactory.CometAPI]: 'cometapi',
  [LLMFactory.DeerAPI]: 'deerapi',
  [LLMFactory.JiekouAI]: 'jiekouai',
  [LLMFactory.Builtin]: 'builtin',
  [LLMFactory.MinerU]: 'mineru',
  [LLMFactory.Perplexity]: 'perplexity',
  [LLMFactory.Gemini]: 'gemini',
};

// API Key URL mapping for each provider
export const APIMapUrl: Partial<Record<LLMFactory, string>> = {
  [LLMFactory.OpenAI]: 'https://platform.openai.com/api-keys',
  [LLMFactory.Anthropic]: 'https://console.anthropic.com/settings/keys',
  [LLMFactory.Gemini]: 'https://aistudio.google.com/app/apikey',
  [LLMFactory.DeepSeek]: 'https://platform.deepseek.com/api_keys',
  [LLMFactory.Moonshot]: 'https://platform.moonshot.cn/console/api-keys',
  [LLMFactory.TongYiQianWen]: 'https://dashscope.console.aliyun.com/apiKey',
  [LLMFactory.ZhipuAI]: 'https://open.bigmodel.cn/usercenter/apikeys',
  [LLMFactory.XAI]: 'https://x.ai/api/',
  [LLMFactory.HuggingFace]: 'https://huggingface.co/settings/tokens',
  [LLMFactory.Mistral]: 'https://console.mistral.ai/api-keys/',
  [LLMFactory.Cohere]: 'https://dashboard.cohere.com/api-keys',
  [LLMFactory.BaiduYiYan]: 'https://wenxin.baidu.com/user/key',
  [LLMFactory.Bedrock]: 'https://us-east-1.console.aws.amazon.com/iam/home?region=us-east-2#/users',
  [LLMFactory.AzureOpenAI]: 'https://portal.azure.com/#create/Microsoft.CognitiveServicesOpenAI',
  [LLMFactory.OpenRouter]: 'https://openrouter.ai/keys',
  [LLMFactory.XunFeiSpark]: 'https://console.xfyun.cn/services/cbm',
  [LLMFactory.MiniMax]: 'https://platform.minimaxi.com/user-center/basic-information',
  [LLMFactory.Groq]: 'https://console.groq.com/keys',
  [LLMFactory.NVIDIA]: 'https://build.nvidia.com/settings/api-keys',
  [LLMFactory.SILICONFLOW]: 'https://cloud.siliconflow.cn/account/ak',
  [LLMFactory.Replicate]: 'https://replicate.com/account/api-tokens',
  [LLMFactory.VolcEngine]: 'https://console.volcengine.com/ark',
  [LLMFactory.TencentHunYuan]: 'https://console.cloud.tencent.com/hunyuan/api-key',
  [LLMFactory.TencentCloud]: 'https://console.cloud.tencent.com/cam/capi',
  [LLMFactory.ModelScope]: 'https://modelscope.cn/my/myaccesstoken',
  [LLMFactory.GoogleCloud]: 'https://console.cloud.google.com/apis/credentials',
  [LLMFactory.FishAudio]: 'https://fish.audio/app/api-keys/',
  [LLMFactory.GiteeAI]: 'https://ai.gitee.com/hhxzgrjn/dashboard/settings/tokens',
  [LLMFactory.StepFun]: 'https://platform.stepfun.com/interface-key',
  [LLMFactory.BaiChuan]: 'https://platform.baichuan-ai.com/console/apikey',
  [LLMFactory.PPIO]: 'https://ppio.com/settings/key-management',
  [LLMFactory.VoyageAI]: 'https://dash.voyageai.com/api-keys',
  [LLMFactory.TogetherAI]: 'https://api.together.xyz/settings/api-keys',
  [LLMFactory.NovitaAI]: 'https://novita.ai/dashboard/key',
  [LLMFactory.Upstage]: 'https://console.upstage.ai/api-keys',
  [LLMFactory.CometAPI]: 'https://api.cometapi.com/console/token',
  [LLMFactory.Ai302]: 'https://302.ai/apis/list',
  [LLMFactory.DeerAPI]: 'https://api.deerapi.com/token',
  [LLMFactory.TokenPony]: 'https://www.tokenpony.cn/#/user/keys',
  [LLMFactory.DeepInfra]: 'https://deepinfra.com/dash/api_keys',
};

// Providers that support custom base URL
export const ModelsWithBaseUrl = [
  LLMFactory.OpenAI,
  LLMFactory.AzureOpenAI,
  LLMFactory.TongYiQianWen,
  LLMFactory.MiniMax,
];

// Providers with special modal implementations
export const SpecialModalProviders = [
  LLMFactory.Bedrock,
  LLMFactory.VolcEngine,
  LLMFactory.TencentHunYuan,
  LLMFactory.XunFeiSpark,
  LLMFactory.BaiduYiYan,
  LLMFactory.FishAudio,
  LLMFactory.TencentCloud,
  LLMFactory.GoogleCloud,
  LLMFactory.AzureOpenAI,
  LLMFactory.MinerU,
];

// Local LLM providers
export const LocalLlmProviders = [
  LLMFactory.Ollama,
  LLMFactory.Xinference,
  LLMFactory.LocalAI,
  LLMFactory.LMStudio,
  LLMFactory.ModelScope,
  LLMFactory.HuggingFace,
  LLMFactory.GPUStack,
  LLMFactory.VLLM,
  LLMFactory.OpenAiAPICompatible,
  LLMFactory.OpenRouter,
  LLMFactory.TogetherAI,
  LLMFactory.Replicate,
  LLMFactory.PPIO,
  LLMFactory.SILICONFLOW,
  LLMFactory.PerfXCloud,
  LLMFactory.NovitaAI,
  LLMFactory.Upstage,
  LLMFactory.LeptonAI,
  LLMFactory.DeepInfra,
  LLMFactory.Ai302,
  LLMFactory.GiteeAI,
  LLMFactory.TokenPony,
  LLMFactory.CometAPI,
  LLMFactory.DeerAPI,
  LLMFactory.JiekouAI,
  LLMFactory.Builtin,
];

export const isLocalLlmFactory = (factory: string): boolean => {
  return LocalLlmProviders.includes(factory as LLMFactory);
};

// Model type constants
export enum LlmModelType {
  Chat = 'chat',
  Embedding = 'embedding',
  Image2text = 'image2text',
  Speech2text = 'speech2text',
  Rerank = 'rerank',
  TTS = 'tts',
  Ocr = 'ocr',
}

// Model type display names
export const ModelTypeLabels: Record<LlmModelType, string> = {
  [LlmModelType.Chat]: 'LLM',
  [LlmModelType.Embedding]: 'TEXT EMBEDDING',
  [LlmModelType.Image2text]: 'IMAGE2TEXT',
  [LlmModelType.Speech2text]: 'SPEECH2TEXT',
  [LlmModelType.Rerank]: 'TEXT RE-RANK',
  [LlmModelType.TTS]: 'TTS',
  [LlmModelType.Ocr]: 'OCR',
};

// Model type order for sorting
export const ModelTypeOrder: Record<string, number> = {
  'LLM': 1,
  'TEXT EMBEDDING': 2,
  'TEXT RE-RANK': 3,
  'TTS': 4,
  'SPEECH2TEXT': 5,
  'IMAGE2TEXT': 6,
  'OCR': 7,
  'MODERATION': 8,
};

// Sort tags based on predefined order
export const sortTags = (tags: string): string[] => {
  return tags
    .split(',')
    .map((tag) => tag.trim())
    .sort((a, b) => (ModelTypeOrder[a] || 999) - (ModelTypeOrder[b] || 999));
};
